usaspending_api/common/retrieve_file_from_uri.py

```diff
@@ -57,6 +57,13 @@ def copy(self, dest_file_path):
         else:
             raise NotImplementedError("No handler for scheme: {}!".format(self.parsed_url_obj.scheme))
 
+    def copy_to_temporary_file(self):
+        """Sometimes it is super helpful to just have a nice, concrete, local file to work with."""
+        with tempfile.NamedTemporaryFile() as tf:
```

Comment:
Does this reliably work?

My understanding of the documentation is after the file is closed (which should occur when exiting the context manager) the file isn't guaranteed to persist. It is up to the garbage collection.

One theory is that `copy_to_temporary_file()` effectively only generates a temp file name (since the temp file is removed immediately), then `self.copy()` re-creates the file when it copies the data into it. If this is the case, does the file always persist? I see os.remove below, so I'm assuming this is what is occurring.

Is that desirable behavior to implement here?

Reply:
This is a very good question and I will have to get back to you... probably not today.

Reply:
There's certainly no shortage of ways to create temporary files in Python.  The goal of this snippet is to get me a file path that is guaranteed to be unique and creatable.  I thought actually creating the file would be a great way to do that, but then I want the file gone because I intend to use it.

A quick walk through `NamedTemporaryFile` seems to indicate that it's a hard delete as soon as the context manager `__exit__` is called.  I also created a little test app that checked for the file immediately after the context manager ends and it was gone.  I don't think it's garbage collected.  Where did you see that?

I'm not necessarily married to this solution.  It was just an example of one way to do it that I had handy.  If you have a more preferred method, I'm all ears.

Reply:
> I don't think it's garbage collected. Where did you see that?

Something in the stdlib documentation page. Might have misread it.